Platform/OpenNtfsDxe/Disc.c

```diff
@@ -0,0 +1,731 @@
+
+  FileSystem->FirstMftRecord = Boot.MftLcn * mClusterSize;
+
+  /* Driver limitations */
+  if ((mFileRecordSize > NTFS_MAX_MFT) || (mIndexRecordSize > NTFS_MAX_IDX)) {
+    DEBUG((DEBUG_INFO, "NTFS: (NtfsMount #4) BIOS Parameter Block is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  RootFile = AllocateZeroPool (sizeof (EFI_NTFS_FILE));
+  if (RootFile == NULL) {
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  CopyMem (RootFile, &FileSystem->EfiFile, sizeof (EFI_FILE_PROTOCOL));
+
+  RootFile->IsDir         = TRUE;
+  RootFile->Path          = L"/";
+  RootFile->RefCount      = 1;
+  RootFile->FileSystem    = FileSystem;
+  RootFile->RootFile.File = RootFile;
+  RootFile->MftFile.File  = RootFile;
+
+  RootFile->MftFile.FileRecord = AllocateZeroPool (mFileRecordSize);
+  if (RootFile->MftFile.FileRecord == NULL) {
+    FreePool (RootFile);
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  Status = DiskRead (
+    FileSystem,
+    FileSystem->FirstMftRecord,
+    mFileRecordSize,
+    RootFile->MftFile.FileRecord
+    );
+  if (EFI_ERROR(Status)) {
+    FreePool (RootFile->MftFile.FileRecord);
+    FreePool (RootFile);
+    return Status;
+  }
+
+  Status = Fixup (RootFile->MftFile.FileRecord, mFileRecordSize, "FILE");
+  if (EFI_ERROR(Status)) {
+    FreePool (RootFile->MftFile.FileRecord);
+    FreePool (RootFile);
+    return Status;
+  }
+
+  if (LocateAttr (&RootFile->MftFile.Attr, &RootFile->MftFile, AT_DATA) == NULL) {
+    FreePool (RootFile->MftFile.FileRecord);
+    FreePool (RootFile);
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  Status = InitFile (&RootFile->RootFile, ROOT_FILE);
+  if (EFI_ERROR(Status)) {
+    FreePool (RootFile->MftFile.FileRecord);
+    FreePool (RootFile);
+    return Status;
+  }
+
+  FileSystem->RootIndex = &RootFile->RootFile;
+  FileSystem->MftStart  = &RootFile->MftFile;
+
+  return EFI_SUCCESS;
+}
+
+/**
+   Table 4.21. Layout of a File Record
+   ____________________________________________________________________
+   Offset | Size | Description
+   --------------------------------------------------------------------
+   0x00   | 4    | Magic number 'FILE'
+   0x04   | 2    | Offset to the update sequence
+   0x06   | 2    | Size in words of Update Sequence Number & Array (S)
+   ...
+          | 2    | Update Sequence Number (a)
+          | 2S-2 | Update Sequence Array (a)
+   --------------------------------------------------------------------
+**/
+EFI_STATUS
+EFIAPI
+Fixup (
+  IN UINT8       *Buffer,
+  IN UINT64      Length,
+  IN CONST CHAR8 *Magic
+  )
+{
+  FILE_RECORD_HEADER *Record;
+  UINT8              *UpdateSequencePointer;
+  UINT16             UpdateSequenceNumber;
+  UINT64             USCounter;
+  UINT8              *BufferEnd;
+
+  Record = (FILE_RECORD_HEADER *) Buffer;
+
+  if (Length < sizeof (FILE_RECORD_HEADER)) {
+    DEBUG((DEBUG_INFO, "NTFS: (Fixup #1) Record is corrupted.\n"));
+    return EFI_VOLUME_CORRUPTED;
+  }
+
+  if (CompareMem (Record->Magic, Magic, 4) != 0) {
```

Comment:
My best guess is that here `4` stands for `L_STR_LEN ("NTFS")`?

Reply:
No, better sizeof(Record->Magic). But I really don’t like passing the magic as a pointer. Can we
1) Have a function that reads the record magic and returns an UINT32? Then the reference magic value can be UINT32 too by using SIGNATURE_32, or
2) Similar to 1), but we just declare the record with a UINT32 magic field? This should not mess with the struct alignment I think.
@vit9696 ?

Reply:
To be honest, going route 2 makes some sense to me. But we may have to mark the structure as packed.